util/sonar.sh

```diff
@@ -2,55 +2,80 @@
-echo "sonar.projectVersion=$version" >> $filename
-echo "sonar.cxx.includeDirectories=$includes" >> $filename
-echo "sonar.password=$SONAR_PASSWORD" >> $filename
-echo "sonar.login=$SONAR_ACCESS_TOKEN" >> $filename
-echo "sonar.host.url=$SONAR_URL" >> $filename
-
-(cd $tmp && cmake -DCMAKE_BUILD_TYPE=Debug \
-                  -DCMAKE_CXX_FLAGS="--coverage -Wall -Wextra -fdiagnostics-show-option" \
-                  -DCMAKE_EXE_LINKER_FLAGS="--coverage" \
-                  -DOGDF_SEPARATE_TESTS=ON \
-                  -DOGDF_WARNING_ERRORS=ON ..) || die "Could not run CMake"
-make -C $tmp -j "$cores" COIN 2>/dev/null || die "make COIN failed"
-make -C $tmp -j "$cores" build-all 2>log/gcc.log || die "make build-all failed"
-util/perform_separate_tests.sh $tmp || die "performing separate tests failed"
-gcovr -x -r . --object-directory=$tmp/CMakeFiles >log/gcovr.xml || die "gcovr failed"
-
-sourcefiles="$(git ls-files 'src/ogdf/*.cpp' 'test/src/*.cpp' | check_filter)"
-unusedheaders="$(util/find_unused_headers.sh $tmp | check_filter)"
-cppcheck -j "$cores" --enable=style $compilerincludes --force --xml --language=c++ $sourcefiles $unusedheaders 2>log/cppcheck.xml || die "cppcheck failed"
-
-# point sonar-scanner to certificate so it can communicate with sonarqube server over https
-sonar_keystore="$(echo /sonar-scanner*/jre/lib/security/cacerts)"
-export SONAR_SCANNER_OPTS=-Djavax.net.ssl.trustStore="$sonar_keystore"
-sonar-scanner || die "sonar-runner failed"
+set -e
+mkdir -p build-sonar sonar
+
+git ls-files --full-name -- 'src/ogdf/*.cpp' 'test/src/*.cpp' | check_filter | sort | uniq > sonar/all-sources.txt
+git ls-files --full-name -- 'include/ogdf/*.h' | check_filter | sort | uniq > sonar/all-headers.txt
+
+## generate compile_commands.json
+opts="cmake .. -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_CXX_COMPILER='clang++' "
+
+## debug config
+opts+="-DCMAKE_CXX_FLAGS_DEBUG='-g -O1' "
+opts+="-DBUILD_SHARED_LIBS=ON "
+opts+="-DCMAKE_BUILD_TYPE=Debug "
+opts+="-DOGDF_DEBUG_MODE=HEAVY "
+opts+="-DOGDF_USE_ASSERT_EXCEPTIONS=ON "
+opts+="-DOGDF_MEMORY_MANAGER=POOL_TS "
+
+## include CGAL
+opts+="-DOGDF_INCLUDE_CGAL=ON "
+opts+="-DCGAL_DO_NOT_WARN_ABOUT_CMAKE_BUILD_TYPE=TRUE "
+
+## coverage
+opts+="-DOGDF_SEPARATE_TESTS=ON "
+opts+="-DCMAKE_CXX_FLAGS='--coverage -Wall -Wextra -fdiagnostics-show-option' "
+opts+="-DCMAKE_EXE_LINKER_FLAGS='--coverage' "
+
+## cmd-line args
+opts+="$@"
+
+# Compile!
+echo "::group::($(date -Iseconds)) Compile"
+cd build-sonar
+echo $opts
+eval $opts
+make -j "$cores" build-all
+cp compile_commands.json ..
+cp compile_commands.json ../sonar
+cat CMakeFiles/OGDF.dir/depend.make \
+  | cut -d ":" -f 2 \
+  | grep "^ ../include/ogdf/" \
+  | $OGDF_XARGS realpath --relative-to=".." -q \
+  | check_filter | sort | uniq \
+  > ../sonar/used-headers.txt
+cd ..
+echo "::endgroup::"
+
+# Check for unused headers
+diff sonar/used-headers.txt sonar/all-headers.txt | sed -ne 's/^> //p' > sonar/unused-headers.txt
```

Comment:
I find `cat sonar/used-headers.txt sonar/all-headers.txt | uniq > sonar/unused-headers.txt` more intuitive but maybe that's just personal taste.

Reply:
Doesn't that do something entirely different? My code prints all lines that are in the first file but not the second, yours prints all lines that are in either file, but filters out adjacent duplicates. Note that the `uniq` docs state:

> Note: 'uniq' does not detect repeated lines unless they are adjacent.
> You may want to sort the input first, or use 'sort -u' without 'uniq'.

Reply:
Forget about it – you're right that I'd need another `sort` for my version to work, and at that point the diff/sed version is at least as good as that.